File: backend/app/providers/cms.py
from __future__ import annotations

from typing import Any, cast

import httpx

from app.core.config import Settings
from app.core.errors import FacilityNotFoundError
from app.models import FacilityData
from app.providers.case_reference import apply_case_reference_snapshot

CLAIMS_MEASURE_LABELS = {
    "521": ("Short Term Hospitalization", "percent"),
    "522": ("STR ED Visit", "percent"),
    "551": ("LT Hospitalization", "decimal"),
    "552": ("ED Visit", "decimal"),
}

AVERAGE_FIELD_LABELS = {
    "percentage_of_short_stay_residents_who_were_rehospitalized__1d02": (
        "STR National Avg. for Hospitalization",
        "STR State National Avg. for Hospitalization",
        "percent",
    ),
    "percentage_of_short_stay_residents_who_had_an_outpatient_em_d911": (
        "STR ED Visits National Avg.",
        "STR ED Visits State Avg.",
        "percent",
    ),
    "number_of_hospitalizations_per_1000_longstay_resident_days": (
        "LT National Avg. for Hospitalization",
        "LT State National Avg. for Hospitalization",
        "decimal",
    ),
    "number_of_outpatient_emergency_department_visits_per_1000_l_de9d": (
        "LT ED Visits National Avg.",
        "LT ED Visits State Avg.",
        "decimal",
    ),
}


class CMSProvider:
    def __init__(self, settings: Settings) -> None:
        self.settings = settings

    async def lookup_facility(self, ccn: str) -> FacilityData:
        async with httpx.AsyncClient(timeout=20) as client:
            provider_row = await self._fetch_provider_row(client, ccn)
            state = self._first(provider_row, "provider_state", "state") or ""
            claims_metrics = await self._fetch_claims_metrics(client, ccn, state)

        facility = self._map_provider(provider_row, ccn, claims_metrics)
        if self.settings.apply_case_reference_snapshot:
            return apply_case_reference_snapshot(facility)
        return facility

    async def _fetch_provider_row(self, client: httpx.AsyncClient, ccn: str) -> dict[str, Any]:
        url = f"{self.settings.cms_api_base_url}/{self.settings.cms_provider_dataset_id}/0"
        response = await client.get(
            url,
            params={
                "conditions[0][property]": "cms_certification_number_ccn",
                "conditions[0][value]": ccn,
                "conditions[0][operator]": "=",
            },
        )
        response.raise_for_status()
        rows = self._rows_from_payload(response.json())
        if not rows:
            raise FacilityNotFoundError(ccn)
        return rows[0]

    async def _fetch_claims_metrics(
        self, client: httpx.AsyncClient, ccn: str, state: str
    ) -> dict[str, str | None]:
        if not self.settings.cms_claims_dataset_id:
            return {}

        claims_rows = await self._fetch_rows(
            client,
            self.settings.cms_claims_dataset_id,
            "cms_certification_number_ccn",
            ccn,
            limit=100,
        )
        state_rows = await self._fetch_rows(
            client,
            self.settings.cms_state_averages_dataset_id,
            "state_or_nation",
            state,
            limit=1,
        )
        nation_rows = await self._fetch_rows(
            client,
            self.settings.cms_state_averages_dataset_id,
            "state_or_nation",
            "NATION",
            limit=1,
        )

        metrics: dict[str, str | None] = {}
        for row in claims_rows:
            code = self._first(row, "measure_code")
            if code not in CLAIMS_MEASURE_LABELS:
                continue
            label, style = CLAIMS_MEASURE_LABELS[code]
            metrics[label] = self._format_metric(self._first(row, "adjusted_score"), style)

        if nation_rows:
            self._merge_average_metrics(metrics, nation_rows[0], is_state=False)
        if state_rows:
            self._merge_average_metrics(metrics, state_rows[0], is_state=True)
        return metrics

    async def _fetch_rows(
        self,
        client: httpx.AsyncClient,
        dataset_id: str,
        property_name: str,
        value: str,
        limit: int,
    ) -> list[dict[str, Any]]:
        url = f"{self.settings.cms_api_base_url}/{dataset_id}/0"
        response = await client.get(
            url,
            params={
                "conditions[0][property]": property_name,
                "conditions[0][value]": value,
                "conditions[0][operator]": "=",
                "limit": str(limit),
            },
        )
        response.raise_for_status()
        return self._rows_from_payload(response.json())

    def _merge_average_metrics(
        self, metrics: dict[str, str | None], row: dict[str, Any], is_state: bool
    ) -> None:
        normalized = {str(key).lower(): value for key, value in row.items()}
        for field, labels in AVERAGE_FIELD_LABELS.items():
            national_label, state_label, style = labels
            value = normalized.get(field)
            if value in (None, ""):
                continue
            label = state_label if is_state else national_label
            metrics[label] = self._format_metric(value, style)

    def _map_provider(
        self,
        row: dict[str, Any],
        ccn: str,
        claims_metrics: dict[str, str | None],
    ) -> FacilityData:
        address = self._first(row, "provider_address", "address")
        city = self._first(row, "provider_city", "citytown", "city")
        state = self._first(row, "provider_state", "state")
        zip_code = self._first(row, "provider_zip_code", "zip_code", "zip")
        location_parts = [address, city, state]
        location = ", ".join(part for part in location_parts if part)
        if zip_code and zip_code not in location:
            location = f"{location} {zip_code}".strip()

        beds = self._first(row, "number_of_certified_beds", "certified_beds")
        return FacilityData(
            ccn=ccn,
            provider_name=self._first(row, "provider_name", "facility_name") or "",
            provider_address=address or "",
            city=city or "",
            state=state or "",
            zip_code=zip_code or "",
            location=location,
            census_capacity=self._int_or_none(beds),
            overall_star_rating=self._string_or_none(self._first(row, "overall_rating")),
            health_inspection_rating=self._string_or_none(
                self._first(row, "health_inspection_rating")
            ),
            staffing_rating=self._string_or_none(self._first(row, "staffing_rating")),
            quality_of_resident_care_rating=self._string_or_none(
                self._first(row, "quality_measure_rating", "qm_rating")
            ),
            claims_metrics=claims_metrics,
            source=row,
        )

    @staticmethod
    def _rows_from_payload(payload: Any) -> list[dict[str, Any]]:
        if isinstance(payload, dict):
            if isinstance(payload.get("results"), list):
                return cast(list[dict[str, Any]], payload["results"])
            if isinstance(payload.get("data"), list):
                return cast(list[dict[str, Any]], payload["data"])
        if isinstance(payload, list):
            return payload
        return []

    @staticmethod
    def _first(row: dict[str, Any], *keys: str) -> str | None:
        lower = {str(key).lower(): value for key, value in row.items()}
        for key in keys:
            value = lower.get(key.lower())
            if value not in (None, ""):
                return str(value).strip()
        return None

    @staticmethod
    def _int_or_none(value: Any) -> int | None:
        try:
            return int(float(str(value)))
        except (TypeError, ValueError):
            return None

    @staticmethod
    def _string_or_none(value: Any) -> str | None:
        if value in (None, ""):
            return None
        return str(value).strip()

    @staticmethod
    def _format_metric(value: Any, style: str = "raw") -> str:
        if value in (None, ""):
            return "N/A"
        if style == "percent":
            return f"{float(str(value)):.1f}%"
        if style == "decimal":
            return f"{float(str(value)):.2f}"
        return str(value).strip()

File: frontend/lib/server/cms.ts
import type { FacilityData } from "@/lib/types";
import { applyCaseReferenceSnapshot } from "./report";

const CMS_API_BASE_URL = "https://data.cms.gov/provider-data/api/1/datastore/query";
const PROVIDER_DATASET_ID = "4pq5-n9py";
const CLAIMS_DATASET_ID = "ijh5-nb2v";
const STATE_AVERAGES_DATASET_ID = "xcdc-v8bm";

const claimsMeasureLabels = {
  "521": ["Short Term Hospitalization", "percent"],
  "522": ["STR ED Visit", "percent"],
  "551": ["LT Hospitalization", "decimal"],
  "552": ["ED Visit", "decimal"],
} as const;

const averageFieldLabels = {
  percentage_of_short_stay_residents_who_were_rehospitalized__1d02: [
    "STR National Avg. for Hospitalization",
    "STR State National Avg. for Hospitalization",
    "percent",
  ],
  percentage_of_short_stay_residents_who_had_an_outpatient_em_d911: [
    "STR ED Visits National Avg.",
    "STR ED Visits State Avg.",
    "percent",
  ],
  number_of_hospitalizations_per_1000_longstay_resident_days: [
    "LT National Avg. for Hospitalization",
    "LT State National Avg. for Hospitalization",
    "decimal",
  ],
  number_of_outpatient_emergency_department_visits_per_1000_l_de9d: [
    "LT ED Visits National Avg.",
    "LT ED Visits State Avg.",
    "decimal",
  ],
} as const;

type CmsRow = Record<string, unknown>;

export async function lookupFacility(ccn: string): Promise<FacilityData> {
  const providerRows = await fetchRows(PROVIDER_DATASET_ID, "cms_certification_number_ccn", ccn, 1);
  const providerRow = providerRows[0];
  if (!providerRow) {
    throw new Error(`No facility found for CCN ${ccn}.`);
  }

  const state = first(providerRow, "provider_state", "state") ?? "";
  const claimsMetrics = await fetchClaimsMetrics(ccn, state);
  return applyCaseReferenceSnapshot(mapProvider(providerRow, ccn, claimsMetrics));
}

async function fetchClaimsMetrics(ccn: string, state: string): Promise<Record<string, string | null>> {
  const [claimsRows, stateRows, nationRows] = await Promise.all([
    fetchRows(CLAIMS_DATASET_ID, "cms_certification_number_ccn", ccn, 100),
    fetchRows(STATE_AVERAGES_DATASET_ID, "state_or_nation", state, 1),
    fetchRows(STATE_AVERAGES_DATASET_ID, "state_or_nation", "NATION", 1),
  ]);

  const metrics: Record<string, string | null> = {};
  for (const row of claimsRows) {
    const code = first(row, "measure_code");
    if (!code || !(code in claimsMeasureLabels)) {
      continue;
    }
    const [label, style] = claimsMeasureLabels[code as keyof typeof claimsMeasureLabels];
    metrics[label] = formatMetric(first(row, "adjusted_score"), style);
  }

  if (nationRows[0]) {
    mergeAverageMetrics(metrics, nationRows[0], false);
  }
  if (stateRows[0]) {
    mergeAverageMetrics(metrics, stateRows[0], true);
  }
  return metrics;
}

async function fetchRows(
  datasetId: string,
  propertyName: string,
  value: string,
  limit: number,
): Promise<CmsRow[]> {
  const params = new URLSearchParams({
    "conditions[0][property]": propertyName,
    "conditions[0][value]": value,
    "conditions[0][operator]": "=",
    limit: String(limit),
  });
  const response = await fetch(`${CMS_API_BASE_URL}/${datasetId}/0?${params.toString()}`, {
    cache: "no-store",
  });
  if (!response.ok) {
    throw new Error("CMS Provider Data Catalog request failed.");
  }
  return rowsFromPayload(await response.json());
}

function mergeAverageMetrics(metrics: Record<string, string | null>, row: CmsRow, isState: boolean) {
  const normalized = normalize(row);
  for (const [field, labels] of Object.entries(averageFieldLabels)) {
    const [nationalLabel, stateLabel, style] = labels;
    const value = normalized[field];
    if (value === null || value === undefined || value === "") {
      continue;
    }
    metrics[isState ? stateLabel : nationalLabel] = formatMetric(value, style);
  }
}

function mapProvider(
  row: CmsRow,
  ccn: string,
  claimsMetrics: Record<string, string | null>,
): FacilityData {
  const address = first(row, "provider_address", "address") ?? "";
  const city = first(row, "provider_city", "citytown", "city") ?? "";
  const state = first(row, "provider_state", "state") ?? "";
  const zipCode = first(row, "provider_zip_code", "zip_code", "zip") ?? "";
  const location = [address, city, state].filter(Boolean).join(", ");
  const beds = first(row, "number_of_certified_beds", "certified_beds");

  return {
    ccn,
    provider_name: first(row, "provider_name", "facility_name") ?? "",
    provider_address: address,
    city,
    state,
    zip_code: zipCode,
    location: zipCode && !location.includes(zipCode) ? `${location} ${zipCode}`.trim() : location,
    census_capacity: intOrNull(beds),
    overall_star_rating: stringOrNull(first(row, "overall_rating")),
    health_inspection_rating: stringOrNull(first(row, "health_inspection_rating")),
    staffing_rating: stringOrNull(first(row, "staffing_rating")),
    quality_of_resident_care_rating: stringOrNull(first(row, "quality_measure_rating", "qm_rating")),
    claims_metrics: claimsMetrics,
  };
}

function rowsFromPayload(payload: unknown): CmsRow[] {
  if (Array.isArray(payload)) {
    return payload as CmsRow[];
  }
  if (payload && typeof payload === "object") {
    const record = payload as Record<string, unknown>;
    if (Array.isArray(record.results)) {
      return record.results as CmsRow[];
    }
    if (Array.isArray(record.data)) {
      return record.data as CmsRow[];
    }
  }
  return [];
}

function first(row: CmsRow, ...keys: string[]): string | null {
  const normalized = normalize(row);
  for (const key of keys) {
    const value = normalized[key.toLowerCase()];
    if (value !== null && value !== undefined && value !== "") {
      return String(value).trim();
    }
  }
  return null;
}

function normalize(row: CmsRow): CmsRow {
  return Object.fromEntries(Object.entries(row).map(([key, value]) => [key.toLowerCase(), value]));
}

function intOrNull(value: unknown): number | null {
  const parsed = Number.parseInt(String(value), 10);
  return Number.isFinite(parsed) ? parsed : null;
}

function stringOrNull(value: unknown): string | null {
  return value === null || value === undefined || value === "" ? null : String(value).trim();
}

function formatMetric(value: unknown, style: "percent" | "decimal"): string {
  if (value === null || value === undefined || value === "") {
    return "N/A";
  }
  const parsed = Number.parseFloat(String(value));
  if (!Number.isFinite(parsed)) {
    return "N/A";
  }
  return style === "percent" ? `${parsed.toFixed(1)}%` : parsed.toFixed(2);
}
